### ISSUE TITLE
SA1633 MalformedFileHeader with "&" in companyName

### ISSUE THREAD

**Issue by nokinger** [DotNetAnalyzers/StyleCopAnalyzers#2145]
Hi, 

in Germany we had often company names like this: "**FooCorp GmbH & Co. KG**"
But the StyleCopyAnalyzer could not handle that, due this line in [FileHeaderHelpers.ParseXmlFileHeader.cs](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers/Helpers/FileHeaderHelpers.cs): 

`var parsedFileHeaderXml = XElement.Parse(xmlString);`

Because the character "&" is illegal in XML elements and "&" will generate an error because the parser interprets it as the start of an character entity.

This means we are not able to set documentationsRules like this:

```
{
  ""settings"": {
    ""documentationRules"": {
      ""companyName"": ""FooCorp GmbH & Co. KG"",
      ""copyrightText"": ""copyright (c) {companyName}. All rights reserved."",
      ""xmlHeader"": false
    }
  }
}
```

How could we solve that?

**Comment by vweijsters** [comment]
XML style headers like the one below will not work indeed and cannot be made to work without changing the analyzer implementation.

```C#
// <copyright file="source.cs" company="FooCorp & BarCorp">
//   copyright (c) FooCorp & BarCorp. All rights reserved.
// </copyright>
```
Regular style headers will work fine (see below for an example)
```C#
// copyright (c) FooCorp & BarCorp. All rights reserved.
```

The settings file you listed above should actually work when you are using the regular style file headers.

**Comment by nokinger** [comment]
sure, the analyzer implementation must be changed. But the question is, would that be option to change the implementation? Because with the old original StyleCop we are able to set the "&" character in xml header!

**Comment by sharwell** [comment]
I'm torn on this. If the header is XML, then the only correct way to add `&` to either of the two locations shown in @vweijsters' XML header example is by writing it as `&amp;`. I would expect that this is  currently possible by using `&amp;` instead of `&` in the stylecop.json configuration.

On the other hand, JSON does not require `&` to be escaped in this way. It seems like the ideal approach is allowing `&` to appear in stylecop.json without escaping it, and then automatically escaping it if (and only if) XML headers are enabled.

On an interesting side note, if #2110 is implemented then @nokinger could get StyleCop Classic behavior by the following:

1. Disable XML headers
2. Set the `copyrightText` property to the complete header in its XML form, but use `{fileName}` and `{companyName}` variables:

    ```
    <copyright file="{fileName}" company="{companyName}">
      copyright (c) {companyName}. All rights reserved.
    </copyright>
    ```

You'd lose some of the specialized warnings relevant to XML headers, but you'd have the ability to include `&` as-is in a non-XML header that just happens to otherwise look like an XML header.

**Comment by nokinger** [comment]
I think i can live with that solution. If also the missing warning are not fully known to me :)

I expect a warning when: 
1. Header is complete missed 
2. When the header not match. 

whichever I do without also on 2.

**Comment by vweijsters** [comment]
With #2110 implemented the mentioned workaround will work properly.
